File: src/plugin/rules/resend/missing-tags.ts
/**
 * resend-missing-tags (integration)
 *
 * Tags power deliverability segmentation in the Resend dashboard. Flags a
 * Resend send whose statically-visible email option object(s) omit a `tags`
 * property. Dynamic argument shapes (e.g. batch built via `.map()`) are skipped
 * to avoid false positives. Advisory only (info).
 */
import { findProperty, getSendOptionObjects, isResendSendCall } from '../../utils/resend.js';

const rule = {
  meta: {
    type: 'suggestion',
    docs: {
      description: 'Resend sends should include tags for deliverability segmentation',
      category: 'integration',
      docsUrl: 'https://resend.com/docs/dashboard/emails/tags',
      recommended: true,
    },
    messages: {
      missingTags:
        'Resend send has no tags. Add tags (e.g. [{ name: "category", value: "welcome" }]) for segmentation.',
    },
    schema: [],
  },
  create(context: any) {
    return {
      CallExpression(node: any) {
        if (!isResendSendCall(node)) return;
        const optionObjects = getSendOptionObjects(node);
        if (optionObjects.length === 0) return; // dynamic shape: cannot tell
        const someMissingTags = optionObjects.some((opts) => !findProperty(opts, 'tags'));
        if (someMissingTags) {
          context.report({ node, messageId: 'missingTags' });
        }
      },
    };
  },
};

export const resendMissingTagsRule = rule;
export default rule;

File: src/providers/resend/rules/python/missing_tags.py
"""resend-missing-tags

Parity with JS: only inspect statically-visible send option dicts. Dynamic
argument shapes (variable payloads) are skipped to avoid false positives.
"""
from __future__ import annotations

import ast
import sys
from pathlib import Path

_PROVIDER = Path(__file__).resolve().parents[2]
if str(_PROVIDER) not in sys.path:
    sys.path.insert(0, str(_PROVIDER))
from utils import dict_get, get_send_option_dicts, is_resend_send

RULE_KEY = "resend-missing-tags"


def check(tree: ast.AST, path: str, source: str) -> list[dict]:
    out = []
    for node in ast.walk(tree):
        if not isinstance(node, ast.Call) or not is_resend_send(node):
            continue
        option_dicts = get_send_option_dicts(node)
        if not option_dicts:
            continue  # dynamic shape: cannot tell
        if any(dict_get(opts, "tags") is None for opts in option_dicts):
            line = getattr(node, "lineno", 1) or 1
            col = (getattr(node, "col_offset", 0) or 0) + 1
            out.append({"file": path, "line": line, "column": col, "ruleKey": RULE_KEY})
    return out
